esmvaltool/diag_scripts/weighting/climwip/calibrate_sigmas.py

```diff
@@ -0,0 +1,367 @@
+        particularly important for a correct calculation of the independence
+        weighting.
+    percentiles : array_like, shape (2,)
+        Lower and upper percentile to use in the confidence test. Has to
+        satisfy 0 <= percentiles <=1 and percentiles[0] < percentiles[1]
+    weighted : bool, optional
+        If weighted is set to False (default: True) all values in the weights
+        matrix will be set to 1, except for nan values which will be preserved.
+        This can be used to calculate the unweighted baseline for each perfect
+        model, consistent with the weighted case.
+
+    Returns
+    -------
+    percentile_spread : array_like, shape (N,)
+        Full range spanned by the two percentiles for each perfect model.
+    inside_ratio: float
+        Ratio of perfect models inside their respective percentile_spread.
+    """
+    percentiles = xr.DataArray(percentiles, dims='percentile')
+
+    # need to rename for the inside_count comparison
+    target_perfect = target.rename(
+        {'model_ensemble': 'perfect_model_ensemble'})
+
+    if not weighted:  # replace with equal weight but keep nans
+        weights_matrix = 0 * weights_matrix + 1
+
+    percentiles_data = xr.apply_ufunc(
+        weighted_quantile,
+        target,
+        percentiles,
+        weights_matrix,
+        input_core_dims=[['model_ensemble'], ['percentile'],
+                         ['model_ensemble']],
+        output_core_dims=[['percentile']],
+        vectorize=True,
+    )
+
+    inside_count = xr.ufuncs.logical_and(
+        target_perfect >= percentiles_data.isel(percentile=0),
+        target_perfect <= percentiles_data.isel(percentile=1)).values
+    inside_ratio = inside_count.sum() / len(inside_count)
+
+    percentiles_spread = (percentiles_data.isel(percentile=1) -
+                          percentiles_data.isel(percentile=0))
+
+    return percentiles_spread, inside_ratio
+
+
+def compute_cost_function(target: 'xr.DataArray',
+                          weights_matrix: 'xr.DataArray',
+                          performance_sigma: float) -> float:
+    """Optimize the performance sigma for confidence.
+
+    Parameters
+    ----------
+    target : array_like, shape (N,)
+        See calculate_percentiles for more information.
+    weights_matrix : array_like, shape (N, N)
+        See calculate_percentiles for more information.
+    performance_sigma : float
+        The performance sigma value used to calculate the weights.
+
+    Returns
+    -------
+    cost_function_value : float
+        Value of the cost function based on the given performance sigma value.
+        The cost function is a discontinuous function distinguishing two cases:
+              99 + abs(difference)  if overconfident
+        f = {
+              sigma                 else
+
+    WARNING
+    -------
+    It is highly recommended to visually inspect the graphical output of this
```

Comment:
Maybe some people will not read this. I think it would be good to put the warning in the readthedocs page as well.

Reply:
Good point! Will do.